src/keys/slip10.rs

```diff
@@ -44,100 +99,402 @@ impl Seed {
+                .map(PublicKey::Secp256k1Ecdsa)
+                .expect("valid extended public key"),
+        }
+    }
+
+    pub fn chain_code(&self) -> &[u8; 32] {
+        self.0.chain_code()
+    }
+
+    pub fn child_key(&self, segment: &Segment) -> crate::Result<Self> {
+        #[cfg(feature = "ed25519")]
+        debug_assert_ne!(Curve::Ed25519, self.curve());
+
+        if !segment.is_hardened() {
+            Ok(Self(self.0.child_key(segment)))
+        } else {
+            Err(crate::Error::InvalidArgumentError {
+                alg: "SLIP10",
+                expected: "non hardened key index for master public key",
+            })
+        }
     }
 
-    pub fn derive(&self, chain: &Chain) -> crate::Result<Key> {
-        let mut k = *self;
-        for c in &chain.0 {
-            k = k.child_key(c)?;
+    pub fn derive(&self, chain: &Chain) -> crate::Result<Self> {
+        #[cfg(feature = "ed25519")]
+        debug_assert_ne!(Curve::Ed25519, self.curve());
+
+        if chain.all_non_hardened() {
+            let mut key = self.0.clone();
+            for segment in &chain.0 {
+                key = key.child_key(segment);
+            }
+            Ok(Self(key))
+        } else {
+            Err(crate::Error::InvalidArgumentError {
+                alg: "SLIP10",
+                expected: "non hardened key index for master public key",
+            })
         }
-        Ok(k)
     }
 }
 
-impl TryFrom<&[u8]> for Key {
+impl TryFrom<&ExtendedSecretKey> for ExtendedPublicKey {
     type Error = crate::Error;
+    fn try_from(esk: &ExtendedSecretKey) -> crate::Result<Self> {
+        esk.try_into_extended_public_key()
+    }
+}
 
-    fn try_from(bs: &[u8]) -> Result<Self, Self::Error> {
-        if bs.len() != 64 {
-            return Err(crate::Error::BufferSize {
-                name: "key",
-                has: bs.len(),
-                needs: 64,
-            });
+#[derive(Clone)]
+struct KeyImpl {
+    curve: Curve,
+    ext: [u8; 65],
+}
+
+impl Zeroize for KeyImpl {
+    fn zeroize(&mut self) {
+        self.ext.zeroize();
+    }
+}
+impl Drop for KeyImpl {
+    fn drop(&mut self) {
+        self.zeroize()
+    }
+}
+impl ZeroizeOnDrop for KeyImpl {}
+
+impl KeyImpl {
+    fn ext_mut(&mut self) -> &mut [u8; 64] {
+        unsafe { &mut *(self.ext[1..].as_mut_ptr() as *mut [u8; 64]) }
+    }
+
+    fn extended_secret_bytes(&self) -> &[u8; 64] {
+        unsafe { &*(self.ext[1..].as_ptr() as *const [u8; 64]) }
+    }
+
+    fn extended_public_bytes(&self) -> &[u8; 65] {
+        &self.ext
+    }
+
+    fn secret_bytes(&self) -> &[u8; 32] {
+        unsafe { &*(self.ext[1..33].as_ptr() as *const [u8; 32]) }
+    }
+
+    fn public_bytes(&self) -> &[u8; 33] {
+        unsafe { &*(self.ext[..33].as_ptr() as *const [u8; 33]) }
+    }
+
+    fn chain_code(&self) -> &[u8; 32] {
+        unsafe { &*(self.ext[33..].as_ptr() as *const [u8; 32]) }
```

Comment:
Maybe also use consts for the 32, 33, 64, 65 numbers here?
`[1..33]` is the first byte a type byte or why is it 33?

Reply:
SLIP-10 is designed for certain curves such as Secp256k1, NIST P-256, Ed25519. These consts are related to curve size, field modulus size, encoded public key size. KeyImpl is a private type, there's no need to name them. Maybe for ExtendedSecretKey and ExtendedPublicKey (extended_bytes, secret_bytes, public_bytes) it makes sense.

The ext[0] byte is there for encoding of a public key using SEC1 format. KeyImpl is used to derive both secret and public keys: for secret key ext[0] == 0 and ext[1..33] encodes a secret key, for public key ext[0..33] contains SEC1 encoding of a public key with ext[0] == 2 or ext[0] == 3. This is a way to implement polymorphism here. Otherwise there would be a copy-paste or some private trait.